cpp: defined popcount macro prevents use of std::popcount

**Describe the bug**

C++20 introduced a new feature called `std::popcount` ([ref](https://en.cppreference.com/w/cpp/numeric/popcount)). However, due to a macro named `popcount` any reference to this new feature is corrupted, because it's instead evaluated to `std::__builtin_popcount`.

https://github.com/zephyrproject-rtos/zephyr/blob/e4fcb32451c587a3e4ba7f8bf3fc602b16f9652b/include/zephyr/toolchain/gcc.h#L253

This results in the following error when attempting to use `std::popcount`:

```
/home/hein/zephyrproject/zephyr/samples/subsys/cpp/cpp_synchronization/src/main.cpp: In function 'int main()':
/home/hein/zephyrproject/zephyr/include/zephyr/toolchain/gcc.h:253:21: error: '__builtin_popcount' is not a member of 'std'; did you mean '__builtin_popcount'?
  253 | #define popcount(x) __builtin_popcount(x)
      |                     ^~~~~~~~~~~~~~~~~~
/home/hein/zephyrproject/zephyr/samples/subsys/cpp/cpp_synchronization/src/main.cpp:146:35: note: in expansion of macro 'popcount'
  146 |                 printk("%d", std::popcount(0xF));
      |                                   ^~~~~~~~
```

**To Reproduce**
Steps to reproduce the behavior:
1. Checkout Zephyr version that supports `STD_CPP20`. (I used https://github.com/zephyrproject-rtos/zephyr/commit/8c8d41c32fb5d7a6a965c4cbdfa955d95905cf4a)
2. Apply the following patch (modifies `sample/sybsus/cpp/cpp_synchronization`)
```patch
diff --git a/samples/subsys/cpp/cpp_synchronization/prj.conf b/samples/subsys/cpp/cpp_synchronization/prj.conf
index a4c3fe1b5c..56041cfc95 100644
--- a/samples/subsys/cpp/cpp_synchronization/prj.conf
+++ b/samples/subsys/cpp/cpp_synchronization/prj.conf
@@ -1,3 +1,5 @@
 CONFIG_CPLUSPLUS=y
 CONFIG_CPP_MAIN=y
 CONFIG_MINIMAL_LIBC_MALLOC_ARENA_SIZE=128
+CONFIG_STD_CPP20=y
+CONFIG_LIB_CPLUSPLUS=y
\ No newline at end of file
diff --git a/samples/subsys/cpp/cpp_synchronization/src/main.cpp b/samples/subsys/cpp/cpp_synchronization/src/main.cpp
index 3ccf39bbb2..cf8793f29d 100644
--- a/samples/subsys/cpp/cpp_synchronization/src/main.cpp
+++ b/samples/subsys/cpp/cpp_synchronization/src/main.cpp
@@ -18,6 +18,7 @@
  * @file C++ Synchronization demo.  Uses basic C++ functionality.
  */
 
+#include <bit>
 #include <stdio.h>
 #include <zephyr/kernel.h>
 #include <zephyr/arch/cpu.h>
@@ -142,6 +143,7 @@ int main(void)
 	while (1) {
 		/* say "hello" */
 		printk("%s: Hello World!\n", __FUNCTION__);
+		printk("%d", std::popcount(0xF));
 
 		/* wait a while, then let coop thread have a turn */
 		k_timer_start(&timer, K_MSEC(SLEEPTIME), K_NO_WAIT);

```
3. `west build -p always -b nucleo_h743zi samples/subsys/cpp/cpp_synchronization`
4. Observe error shown earlier.

**Expected behavior**
For `std::popcount` to be compile and function normally.

**Impact**
Our project relies https://www.etlcpp.com/ which utilizes these new features if they're available. ~~As workaround we will likely have to build this library now with `ETL_NO_STL`. The symptoms of this workaround isn't investigated yet, but might reduce performance and/or increase code size. So for the moment the impact is contained, but would prefer it fixed.~~ 

**Update:** This is a **blocking** problem with using ETL. See first comment.

**Environment (please complete the following information):**
 - OS: ubuntu 22.04
 - Toolchain: zephyr SDK 0.15.2
 - Commit: https://github.com/zephyrproject-rtos/zephyr/commit/8c8d41c32fb5d7a6a965c4cbdfa955d95905cf4a

Update: this is a _blocking_ issue which prevents the use of https://www.etlcpp.com/ with C++20. It's not possible to build ETL with `ETL_NO_STL` with an actual `STL` present. It causes redefinition errors, as one might expect.
```
/home/hein/redacted/vendor/etlcpp/src/etlcpp-external/include/etl/initializer_list.h:150:9: error: redefinition of 'class std::initializer_list<_E>'
  150 |   class initializer_list
      |         ^~~~~~~~~~~~~~~~
In file included from /home/hein/zephyr-sdk-0.15.2/arm-zephyr-eabi/arm-zephyr-eabi/include/c++/12.1.0/array:39,
                 from /home/hein/redacted.hh:3:
/home/hein/zephyr-sdk-0.15.2/arm-zephyr-eabi/arm-zephyr-eabi/include/c++/12.1.0/initializer_list:47:11: note: previous definition of 'class std::initializer_list<_E>'
   47 |     class initializer_list
      |           ^~~~~~~~~~~~~~~~
```

--- a/drivers/dai/intel/alh/alh.c
+++ b/drivers/dai/intel/alh/alh.c
@@ -50,7 +50,7 @@ static int dai_alh_set_config_blob(struct dai_intel_alh *dp, const struct dai_co
 
 		if (IPC4_ALH_DAI_INDEX(alh_id) == dp->index) {
 			alh->params.stream_id = alh_id;
-			alh->params.channels = popcount(alh_cfg->mapping[i].channel_mask);
+			alh->params.channels = POPCOUNT(alh_cfg->mapping[i].channel_mask);
 			break;
 		}
 	}

--- a/drivers/dai/intel/dmic/dmic_nhlt.c
+++ b/drivers/dai/intel/dmic/dmic_nhlt.c
@@ -253,7 +253,7 @@ int dai_dmic_set_config_nhlt(struct dai_intel_dmic *dmic, const void *bespoke_cf
 
 	/* Channel_ctlr_mask bits indicate the FIFOs enabled*/
 	channel_ctrl_mask = ((struct nhlt_dmic_channel_ctrl_mask *)p)->channel_ctrl_mask;
-	num_fifos = popcount(channel_ctrl_mask); /* Count set bits */
+	num_fifos = POPCOUNT(channel_ctrl_mask); /* Count set bits */
 	p += sizeof(struct nhlt_dmic_channel_ctrl_mask);
 	LOG_DBG("dmic_set_config_nhlt(): channel_ctrl_mask = %d", channel_ctrl_mask);
 
@@ -336,7 +336,7 @@ int dai_dmic_set_config_nhlt(struct dai_intel_dmic *dmic, const void *bespoke_cf
 
 	/* Get PDMx registers */
 	pdm_ctrl_mask = ((struct nhlt_pdm_ctrl_mask *)p)->pdm_ctrl_mask;
-	num_pdm = popcount(pdm_ctrl_mask); /* Count set bits */
+	num_pdm = POPCOUNT(pdm_ctrl_mask); /* Count set bits */
 	p += sizeof(struct nhlt_pdm_ctrl_mask);
 	LOG_DBG("dmic_set_config_nhlt(): pdm_ctrl_mask = %d", pdm_ctrl_mask);
 	if (num_pdm < 1 || num_pdm > CONFIG_DAI_DMIC_HW_CONTROLLERS) {

--- a/drivers/dai/intel/ssp/ssp.c
+++ b/drivers/dai/intel/ssp/ssp.c
@@ -1348,8 +1348,8 @@ static int dai_ssp_set_config_tplg(struct dai_intel_ssp *dp, const struct dai_co
 		 */
 		sspsp |= SSPSP_SFRMP(!inverted_frame);
 
-		active_tx_slots = popcount(ssp->params.tx_slots);
-		active_rx_slots = popcount(ssp->params.rx_slots);
+		active_tx_slots = POPCOUNT(ssp->params.tx_slots);
+		active_rx_slots = POPCOUNT(ssp->params.rx_slots);
 
 		/*
 		 * handle TDM mode, TDM mode has padding at the end of
@@ -1837,9 +1837,9 @@ static const struct dai_config *dai_ssp_config_get(const struct device *dev, enu
 	params->rate = ssp->params.fsync_rate;
 
 	if (dir == DAI_DIR_PLAYBACK) {
-		params->channels = popcount(ssp->params.tx_slots);
+		params->channels = POPCOUNT(ssp->params.tx_slots);
 	} else {
-		params->channels = popcount(ssp->params.rx_slots);
+		params->channels = POPCOUNT(ssp->params.rx_slots);
 	}
 
 	params->word_size = ssp->params.sample_valid_bits;

--- a/include/zephyr/toolchain/gcc.h
+++ b/include/zephyr/toolchain/gcc.h
@@ -250,7 +250,7 @@ do {                                                                    \
 
 #define likely(x)   (__builtin_expect((bool)!!(x), true) != 0L)
 #define unlikely(x) (__builtin_expect((bool)!!(x), false) != 0L)
-#define popcount(x) __builtin_popcount(x)
+#define POPCOUNT(x) __builtin_popcount(x)
 
 #ifndef __no_optimization
 #define __no_optimization __attribute__((optimize("-O0")))

--- a/subsys/bluetooth/mesh/lpn.c
+++ b/subsys/bluetooth/mesh/lpn.c
@@ -748,10 +748,10 @@ static inline int group_popcount(atomic_t *target)
 	int i, count = 0;
 
 	for (i = 0; i < ARRAY_SIZE(bt_mesh.lpn.added); i++) {
-		count += popcount(atomic_get(&target[i]));
+		count += POPCOUNT(atomic_get(&target[i]));
 	}
 #else
-	return popcount(atomic_get(target));
+	return POPCOUNT(atomic_get(target));
 #endif
 }
 

--- a/subsys/bluetooth/mesh/transport.c
+++ b/subsys/bluetooth/mesh/transport.c
@@ -1002,7 +1002,7 @@ static inline int32_t ack_timeout(struct seg_rx *rx)
 	     (ttl * (int32_t)CONFIG_BT_MESH_SEG_ACK_PER_HOP_TIMEOUT);
 
 	/* Add timeout for evenry not yet received segment. */
-	to += ((rx->seg_n + 1) - popcount(rx->block)) *
+	to += ((rx->seg_n + 1) - POPCOUNT(rx->block)) *
 		(int32_t)CONFIG_BT_MESH_SEG_ACK_PER_SEGMENT_TIMEOUT;
 
 	/* Make sure we don't send more frequently than the duration for

--- a/tests/kernel/mem_protect/userspace/src/main.c
+++ b/tests/kernel/mem_protect/userspace/src/main.c
@@ -878,7 +878,7 @@ ZTEST(userspace, test_object_recycle)
 		     "object wasn't marked as initialized");
 
 	for (int i = 0; i < CONFIG_MAX_THREAD_BYTES; i++) {
-		perms_count += popcount(ko->perms[i]);
+		perms_count += POPCOUNT(ko->perms[i]);
 	}
 
 	zassert_true(perms_count == 1, "invalid number of thread permissions");

